### PR TITLE
TS-4699: Adds a C++ API WebSocket example

### PR DESCRIPTION
**Pull request by ogoodman**
Implements a simple WebSocket echo server as a C++ API InterceptPlugin.

### CODE REVIEW

File: lib/atscppapi/examples/experimental/websocket/WebSocket.cc [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+// DISCLAIMER: this is intended for demonstration purposes only and
+// does not pretend to implement a complete (or useful) server.
+
+using namespace atscppapi;
+
+void
+TSPluginInit(int argc, const char *argv[])
+{
+  RegisterGlobalPlugin("WebSocket", "Apache", "support@example.com");
```
Comment (jpeach) [inline lib/atscppapi/examples/experimental/websocket/WebSocket.cc]: Make this consistent with other C++ API example plugins.

File: lib/atscppapi/examples/experimental/websocket/WSBuffer.hh [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#ifndef WS_BUFFER_HH_HEADERS_
+#define WS_BUFFER_HH_HEADERS_
```
Comment (jpeach) [inline lib/atscppapi/examples/experimental/websocket/WSBuffer.hh]: Use ``./tools/mkheader`` to generate a header template, which will include the necessary ASL headers.

We don't use ``.hh`` as a file extension, this should be ``WSBuffer.h``.

File: lib/atscppapi/examples/experimental/websocket/WSBuffer.cc [hunk 3]
```diff
@@ -0,0 +1,206 @@
+#include "WSBuffer.hh"
```
Comment (jpeach) [inline lib/atscppapi/examples/experimental/websocket/WSBuffer.cc]: Needs ASL license header.

File: lib/atscppapi/examples/Makefile.am [hunk 4]
```diff
@@ -15,6 +15,8 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
+SUBDIRS = experimental
```
Comment (jpeach) [inline lib/atscppapi/examples/Makefile.am]: I don't think that we need experimental example plugins. Just make this an example.

File: lib/atscppapi/examples/experimental/websocket/WSBuffer.cc [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+  }
+  EVP_MD_CTX_cleanup(&digest);
+  if (hash_len != 20) {
+    return "bad-hash-length";
+  }
+
+  char digest_buf[WS_DIGEST_MAX];
+  size_t digest_len = 0;
+
+  ats_base64_encode(hash_buf, hash_len, digest_buf, WS_DIGEST_MAX, &digest_len);
```
Comment (jpeach) [inline lib/atscppapi/examples/experimental/websocket/WSBuffer.cc]: This is an example plugin, so you should avoid internal libraries. Use ``TSBase64Encode``.

File: lib/atscppapi/examples/Makefile.am (hunk 4 above)
Comment (ogoodman) [inline lib/atscppapi/examples/Makefile.am]: Ok. Originally I made it just an example and Brian Geffon said it should go in an experimental subdirectory. I'm happy to move it back if that's now the preferred option.

Comment (ogoodman) [inline lib/atscppapi/examples/Makefile.am]: Moved back into examples and squashed.

**Comment by jpeach** [comment]
@ogoodman Would you mind moving the C++ API changes to the other PR so we can keep API changes together?

**Comment by ogoodman** [comment]
Done (API changes consolidated).

**Comment by ogoodman** [comment]
I've added some '#includes' and '#defines' to ensure that be64toh is available on what is hopefully a superset of all supported platforms, based on https://gist.github.com/panzi/6856583. It (the portable be64toh code) works on Ubuntu 14.04 and Mac OS 10.11.6 at least.

**Comment by jpeach** [comment]
[approve ci]

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/867/ for details.

**Comment by jpeach** [comment]
FreeBSD build:
```
../../../../lib/atscppapi/examples/websocket/WSBuffer.cc:40:9: error: 'be64toh' macro redefined [-Werror]
#define be64toh(x) betoh64(x)
        ^
/usr/include/sys/endian.h:77:9: note: previous definition is here
#define be64toh(x)      bswap64((x))
        ^
../../../../lib/atscppapi/examples/websocket/WSBuffer.cc:110:15: error: use of undeclared identifier 'betoh64'
    msg_len = be64toh(*(uint64_t *)(ws_buf_.data() + 2));
              ^
../../../../lib/atscppapi/examples/websocket/WSBuffer.cc:40:20: note: expanded from macro 'be64toh'
#define be64toh(x) betoh64(x)
                   ^
2 errors generated.
```

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/763/ for details.

**Comment by ogoodman** [comment]
Ok, this time I had a look at the source for the relevant header files of each distribution. It seems you can't trust randomly posted, heavily upvoted gists claiming to do useful things. Who would have thought it?

**Comment by jpeach** [comment]
[approve ci]

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/873/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/769/ for details.

**Comment by PSUdaemon** [comment]
@ogoodman the commit message should have the JIRA issue in it, for future reference. Just like the title of this PR.

**Pull request merged**